src/slang-nodes/index.d.ts

```diff
@@ -0,0 +1,474 @@
+  | NamedArgumentsDeclaration
+  | NamedArgumentGroup
+  | NamedArgument
+  | TypeExpression
+  | NewExpression
+  | TupleExpression
+  | TupleValue
+  | ArrayExpression
+  | HexNumberExpression
+  | DecimalNumberExpression
+  | YulBlock
+  | YulFunctionDefinition
+  | YulParametersDeclaration
+  | YulReturnsDeclaration
+  | YulVariableDeclarationStatement
+  | YulVariableDeclarationValue
+  | YulVariableAssignmentStatement
+  | YulStackAssignmentStatement
+  | YulStackAssignmentOperator
+  | YulColonAndEqual
+  | YulEqualAndColon
+  | YulIfStatement
+  | YulForStatement
+  | YulSwitchStatement
+  | YulDefaultCase
+  | YulValueCase
+  | YulLeaveStatement
+  | YulBreakStatement
+  | YulContinueStatement
+  | YulLabel
+  | YulFunctionCallExpression
+  | SourceUnitMember
+  | Pragma
+  | ExperimentalFeature
+  | VersionExpression
+  | ImportClause
+  | UsingClause
+  | UsingOperator
+  | UsingTarget
+  | ContractMember
+  | StateVariableAttribute
+  | FunctionName
+  | FunctionAttribute
+  | FunctionBody
+  | ConstructorAttribute
+  | UnnamedFunctionAttribute
+  | FallbackFunctionAttribute
+  | ReceiveFunctionAttribute
+  | ModifierAttribute
+  | TypeName
+  | FunctionTypeAttribute
+  | MappingKeyType
+  | ElementaryType
+  | Statement
+  | TupleMember
+  | VariableDeclarationType
+  | StorageLocation
+  | ForStatementInitialization
+  | ForStatementCondition
+  | Expression
+  | ArgumentsDeclaration
+  | NumberUnit
+  | StringExpression
+  | StringLiteral
+  | HexStringLiteral
+  | UnicodeStringLiteral
+  | YulStatement
+  | YulAssignmentOperator
+  | YulSwitchCase
+  | YulExpression
+  | YulPathComponent
+  | YulBuiltInFunction
+  | YulLiteral
+  | SourceUnitMembers
+  | VersionExpressionSet
+  | ContractMembers
+  | InterfaceMembers
+  | LibraryMembers
+  | StructMembers
+  | StateVariableAttributes
+  | FunctionAttributes
+  | ConstructorAttributes
+  | UnnamedFunctionAttributes
+  | FallbackFunctionAttributes
+  | ReceiveFunctionAttributes
+  | ModifierAttributes
+  | FunctionTypeAttributes
+  | Statements
+  | CatchClauses
+  | StringLiterals
+  | HexStringLiterals
+  | UnicodeStringLiterals
+  | YulStatements
+  | YulSwitchCases
+  | VersionExpressionSets
+  | VersionSpecifiers
+  | ImportDeconstructionSymbols
+  | UsingDeconstructionSymbols
+  | InheritanceTypes
+  | EnumMembers
+  | Parameters
+  | OverridePaths
+  | EventParameters
+  | ErrorParameters
+  | AssemblyFlags
+  | TupleDeconstructionElements
+  | PositionalArguments
+  | NamedArguments
+  | CallOptions
+  | TupleValues
+  | ArrayValues
+  | IdentifierPath
+  | YulParameters
+  | YulVariableNames
+  | YulArguments
+  | YulPaths
+  | YulPath;
+
+export type CollectionNode = Extract<StrictAstNode, { items: unknown[] }>;
+
+export type BinaryOperation = Extract<
+  StrictAstNode,
+  {
+    leftOperand: Expression;
+    operator: string;
+    rightOperand: Expression;
+  }
+>;
+
+export type AstNode =
+  | StrictAstNode
+  | Comment
+  | Identifier
+  | YulIdentifier
```

Comment:
@OmarTawfik It would be beneficial if `Identifier` and `YulIdentifier` were in the types.
For example:
```TypeScript
export declare class EnumDefinition {
    readonly cst: NonterminalNode;
    private readonly fetch;
    constructor(cst: NonterminalNode);
    get enumKeyword(): TerminalNode;
    get name(): TerminalNode;
    get openBrace(): TerminalNode;
    get members(): EnumMembers;
    get closeBrace(): TerminalNode;
}
```
I care about getting the location information of `name` but not from `enumKeyword`, `openBrace` and `closeBrace`.
I went through every `TerminalNode` in the AST and plucked out the `Identifier`s and `YulIdentifier`s but if you make a change in the AST I have no Type security that will trigger an error on my side if I need to adapt. So I would appreciate if it would be possible to deliver something like:

```TypeScript
export declare class EnumDefinition {
    readonly cst: NonterminalNode;
    private readonly fetch;
    constructor(cst: NonterminalNode);
    get enumKeyword(): TerminalNode;
    get name(): Identifier;
    get openBrace(): TerminalNode;
    get members(): EnumMembers;
    get closeBrace(): TerminalNode;
}
```

Reply:
Thanks for the suggestion!

> but if you make a change in the AST I have no Type security that will trigger an error on my side

To make sure I understand you correctly, do you mean `getNodeMetadata().offsets`?
Is that because you are storing `offsets` as an array, and accessing it by a numeric index? Would it help if it was changed to be an object/dict with the field name as keys? for example `offsets.name` here won't break if the position of `name` was changed relative to its siblings.

> It would be beneficial if Identifier and YulIdentifier were in the types.

How would that help in this case? if they were non-terminal nodes? how would this be tracked otherwise?

One other way you can use Slang API if you want, so that you don't have to do the manual tracking of offsets yourself, is to use cursors.
For example, in `getNodeMetadata()`, instead of iterating over `ast.cst.children()`, you can `ast.cst.createCursor()`, and then use `cursor.goToNextSibling()` to iterate, and `cursor.textOffset()` or `cursor.textRange()` will be updated.

Reply:
🤔 using a Map with the cst as the key could really improve this aspect of the implementation.
it would even render the passing of an offset moot since the we are already passing the AST to the constructors.
I'll experiment with this and see how it behaves.

Reply:
Thanks for pointing out. The `id` already exists in the Rust API, and I would be happy to add it in TypeScript. It should be a simple change.
When I'm done with WASM, I will also be able to expose the field names themselves as a strongly-typed `EdgeLabel` enum, so you would be able to refer to and persist the name of `get name(): Identifier ` field as `EdgeLabel.Name`.